pkg/gator/expand.go

```diff
@@ -0,0 +1,225 @@
+}
+
+func Expand(resources []*unstructured.Unstructured) ([]*unstructured.Unstructured, error) {
+	mutSystem := mutation.NewSystem(mutation.SystemOpts{})
+	expSystem := expansion.NewSystem()
+	er := expansionResources{}
+	if err := er.addResources(resources); err != nil {
+		return nil, fmt.Errorf("error parsing resources: %s", err)
+	}
+
+	for _, te := range er.templateExpansions {
+		if err := expSystem.UpsertTemplate(te); err != nil {
+			return nil, err
+		}
+	}
+
+	var resultants []*unstructured.Unstructured
+	for _, gen := range er.generators {
+		ns, err := er.namespaceForGenerator(gen)
+		base := &types.Mutable{
+			Object:    gen,
+			Namespace: ns,
+			Username:  "",
+			Source:    types.SourceTypeGenerated,
+		}
+		if err != nil {
+			return nil, fmt.Errorf("error expanding generator: %s", err)
+		}
+		r, err := expSystem.Expand(base, mutSystem)
+		if err != nil {
+			return nil, fmt.Errorf("error expanding generator: %s", err)
+		}
+		resultants = append(resultants, r...)
+	}
+
+	return resultants, nil
+}
+
+func (er *expansionResources) namespaceForGenerator(gen *unstructured.Unstructured) (*corev1.Namespace, error) {
+	genNs := gen.GetNamespace()
+	if genNs == "" {
+		return &corev1.Namespace{}, nil
+	}
+	if genNs == "default" {
+		return &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: "default"}}, nil
+	}
+
+	ns, exists := er.namespaces[genNs]
+	if !exists {
+		return nil, fmt.Errorf("namespace resource %q not found in supplied configs", genNs)
+	}
+	return ns, nil
+}
+
+func (er *expansionResources) addResources(resources []*unstructured.Unstructured) error {
+	for _, r := range resources {
+		if err := er.add(r); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (er *expansionResources) addMutator(mut *unstructured.Unstructured) error {
+	var mutErr error
+	var m types.Mutator
+
+	switch mut.GetKind() {
+	case "Assign":
+		a, err := convertAssign(mut)
+		if err != nil {
+			return err
+		}
+		m, mutErr = assign.MutatorForAssign(a)
+	case "AssignMetadata":
+		a, err := convertAssignMetadata(mut)
+		if err != nil {
+			return err
+		}
+		m, mutErr = assignmeta.MutatorForAssignMetadata(a)
+	case "ModifySet":
+		ms, err := convertModifySet(mut)
+		if err != nil {
+			return err
+		}
+		m, mutErr = modifyset.MutatorForModifySet(ms)
+	default:
+		return fmt.Errorf("cannot convert mutator of kind %q", mut.GetKind())
+	}
+
+	if mutErr != nil {
+		return mutErr
+	}
+	if m.Source() != types.SourceTypeAll || m.Source() != types.SourceTypeGenerated {
+		return fmt.Errorf("mutator %s must have 'source: All|Generated'", m.ID().Name)
+	}
+	er.mutators = append(er.mutators, m)
+	return nil
+}
+
+func (er *expansionResources) add(u *unstructured.Unstructured) error {
+	var err error
+	switch {
+	case isMutator(u):
+		err = er.addMutator(u)
+	case isExpansion(u):
+		err = er.addTemplateExpansion(u)
+	case u.GetKind() == "Namespace":
+		err = er.addNamespace(u)
+	default:
+		er.generators = append(er.generators, u)
```

Comment:
Everything is a potential generator, so er.generators should include namespaces, template expansions, and mutators (i.e. don't make this statement part of the `switch`